kuma/javascript/src/payments/subscription-form.jsx

```diff
@@ -1,76 +1,183 @@
+        let script;
+        if (!loadingPromise) {
+            script = document.createElement('script');
+            setLoadingPromise(
+                new Promise((resolve, reject) => {
+                    script.onload = resolve;
+                    script.onerror = reject;
+                })
+            );
+            script.src = url;
+
+            if (document.head) {
+                document.head.appendChild(script);
+            }
+        }
+        return () => {
+            if (document.head && script) {
+                document.head.removeChild(script);
+            }
+        };
+    }, [loadingPromise, url]);
+
+    return [loadingPromise, () => setLoadingPromise(null)];
+}
 
 export default function SubscriptionForm() {
     const locale = getLocale();
-    const subscriptionFormRef = useRef(null);
-    const [isSubmitting, setIsSubmitting] = useState(false);
+    const userData = useContext(UserProvider.context);
+
     const [paymentAuthorized, setPaymentAuthorized] = useState(false);
+    const [formStep, setFormStep] = useState<
+        'initial' | 'stripe_error' | 'stripe' | 'submitting' | 'server_error'
+    >('initial');
 
-    const toggleButton = () => {
-        setPaymentAuthorized(!paymentAuthorized);
-    };
+    const token = useRef(null);
 
-    /**
-     * Opens Stripe modal allowing a user to complete their subscription.
-     * @param {Object} event - The form submit event
-     */
-    const submit = (event) => {
-        event.preventDefault();
+    const [stripeLoadingPromise, reloadStripe] = useScriptLoading(
+        'https://checkout.stripe.com/checkout.js'
+    );
 
-        const subscriptionForm = subscriptionFormRef.current;
-        if (subscriptionForm) {
-            const formData = new FormData(subscriptionForm);
-            setIsSubmitting(true);
+    useEffect(() => {
+        if (!stripeLoadingPromise) {
+            return;
+        }
+        stripeLoadingPromise
+            .then(() => {
+                if (formStep === 'stripe_error') {
+                    setFormStep('initial');
+                }
+            })
+            .catch(() => {
+                setFormStep('stripe_error');
+            });
+    }, [formStep, stripeLoadingPromise]);
 
+    function openStripeModal() {
+        if (!stripeLoadingPromise) {
+            return;
+        }
+        setFormStep('stripe');
+        stripeLoadingPromise.then(() => {
             const stripeHandler = window.StripeCheckout.configure({
                 key: window.mdn.stripePublicKey,
-                locale: 'auto',
+                locale,
                 name: 'MDN Web Docs',
                 zipCode: true,
                 currency: 'usd',
                 amount: 500,
-                email: '',
-                token: function (response) {
-                    formData.set('stripe_token', response.id);
-                    subscriptionForm.submit();
+                email: userData ? userData.email : '',
```

Comment:
How can you *not* have a `userData` at this point?

Reply:
you can't, but the type system does not know that unfortunately.